## Web scraping images of book cover

In [2]:
import os
import codecs
import pandas as pd
from argparse import ArgumentParser
from urllib import request
from tqdm import trange
from joblib import Parallel, delayed
import io

#### The dataset csv can be found here:
https://raw.githubusercontent.com/uchidalab/book-dataset/master/Task1/book30-listing-train.csv
#### Images of book cover are downloaded from Amazon.com

In [3]:
# Read the csv file
header_names = ['Amazon ID (ASIN)', 'Filename', 'Image URL', 'Title', 'Author', 'Category ID',
                'Category']
#with codecs.open(Book_train.csv, mode='r', encoding='utf-8', errors='ignore') as f:
df = pd.read_csv('book_train.csv',delimiter=",", header=None, names=header_names)

In [4]:
df.head()

Amazon ID (ASIN)        Filename  \
0       1588345297  1588345297.jpg   
1       1404803335  1404803335.jpg   
2       1446276082  1446276082.jpg   
3       1491522666  1491522666.jpg   
4        970096410  0970096410.jpg   

                                           Image URL  \
0  http://ecx.images-amazon.com/images/I/51l6XIoa...   
1  http://ecx.images-amazon.com/images/I/51UJnL3T...   
2  http://ecx.images-amazon.com/images/I/51MGUKhk...   
3  http://ecx.images-amazon.com/images/I/51qKvjsi...   
4  http://ecx.images-amazon.com/images/I/51qoUENb...   

                                               Title               Author  \
0         With Schwarzkopf: Life Lessons of The Bear              Gus Lee   
1  Magnets: Pulling Together, Pushing Apart (Amaz...  Natalie M. Rosinsky   
2  Energy Security (SAGE Library of International...                   10   
3       An Amish Gathering: Life in Lancaster County         Beth Wiseman   
4  City of Rocks Idaho: A Climber's Guide (Region...         Dave Bingham   

                    Category ID                  Category  
0                             1     Biographies & Memoirs  
1                             4          Children's Books  
2  Engineering & Transportation                       NaN  
3                             9  Christian Books & Bibles  
4                            26         Sports & Outdoors

In [5]:
df.shape

(51300, 7)

In [ ]:
# This dataset contains information about 51300 images
# Removing the Null values from above dataset 

In [13]:
df.dropna(inplace=True)
df.shape

(47863, 7)

In [12]:
df.groupby(['Category']).count()

Amazon ID (ASIN)  Filename  Image URL  Title  \
Category                                                                     
Arts & Photography                        1487      1487       1487   1487   
Biographies & Memoirs                     1679      1679       1679   1679   
Business & Money                          1629      1629       1629   1629   
Calendars                                 1388      1388       1388   1388   
Children's Books                          1653      1653       1653   1653   
Christian Books & Bibles                  1579      1579       1579   1579   
Comics & Graphic Novels                   1617      1617       1617   1617   
Computers & Technology                    1633      1633       1633   1633   
Cookbooks, Food & Wine                    1651      1651       1651   1651   
Crafts, Hobbies & Home                    1594      1594       1594   1594   
Engineering & Transportation              1593      1593       1593   1593   
Health, Fitness & Dieting                 1586      1586       1586   1586   
History                                   1598      1598       1598   1598   
Humor & Entertainment                     1584      1584       1584   1584   
Law                                       1562      1562       1562   1562   
Literature & Fiction                      1588      1588       1588   1588   
Medical Books                             1442      1442       1442   1442   
Mystery, Thriller & Suspense              1670      1670       1670   1670   
Parenting & Relationships                 1670      1670       1670   1670   
Politics & Social Sciences                1556      1556       1556   1556   
Reference                                 1547      1547       1547   1547   
Religion & Spirituality                   1577      1577       1577   1577   
Romance                                   1673      1673       1673   1673   
Science & Math                            1532      1532       1532   1532   
Science Fiction & Fantasy                 1609      1609       1609   1609   
Self-Help                                 1668      1668       1668   1668   
Sports & Outdoors                         1630      1630       1630   1630   
Teen & Young Adult                        1622      1622       1622   1622   
Test Preparation                          1681      1681       1681   1681   
Travel                                    1565      1565       1565   1565   

                              Author  Category ID  
Category                                           
Arts & Photography              1487         1487  
Biographies & Memoirs           1679         1679  
Business & Money                1629         1629  
Calendars                       1388         1388  
Children's Books                1653         1653  
Christian Books & Bibles        1579         1579  
Comics & Graphic Novels         1617         1617  
Computers & Technology          1633         1633  
Cookbooks, Food & Wine          1651         1651  
Crafts, Hobbies & Home          1594         1594  
Engineering & Transportation    1593         1593  
Health, Fitness & Dieting       1586         1586  
History                         1598         1598  
Humor & Entertainment           1584         1584  
Law                             1562         1562  
Literature & Fiction            1588         1588  
Medical Books                   1442         1442  
Mystery, Thriller & Suspense    1670         1670  
Parenting & Relationships       1670         1670  
Politics & Social Sciences      1556         1556  
Reference                       1547         1547  
Religion & Spirituality         1577         1577  
Romance                         1673         1673  
Science & Math                  1532         1532  
Science Fiction & Fantasy       1609         1609  
Self-Help                       1668         1668  
Sports & Outdoors               1630         1630  
Teen & Young Adult       

### So there are total 47863 total images and 20 different categories are available

In [ ]:
# Function to download image from Amazon.com
def download_image(i):
    filename = df.iloc[i]['Filename']
    category = df.iloc[i]['Category']
    print(category)
    inner_output_dirpath = os.path.join('Desktop\BookDataset', category)
    
    if not os.path.isdir(inner_output_dirpath):
        os.mkdir(inner_output_dirpath)
    output_filepath = os.path.join(inner_output_dirpath, filename)

    url = df.iloc[i]['Image URL']
    if not os.path.isfile(output_filepath):
        downloaded_img = request.urlopen(url)
        f = open(output_filepath, mode='wb')
        f.write(downloaded_img.read())
        downloaded_img.close()
        f.close()

Parallel(n_jobs=-1)(delayed(download_image)(i) for i in trange(len(df)))

100%|██████████| 47863/47863 [2:22:00<00:00,  5.62it/s]      
